modules/mqtt/source/mqtt-source.c

```diff
@@ -0,0 +1,288 @@
+                log_pipe_location_tag(&self->super.super.super.super.super));
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static gint
+_log_ssl_errors(const gchar *str, gsize len, gpointer u)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *) u;
+
+  msg_error("MQTT TLS error", evt_tag_printf("line", "%.*s", (gint) len, str),
+            log_pipe_location_tag(&self->super.super.super.super.super));
+  return TRUE;
+}
+
+static gboolean
+_subscribe_topic(MQTTSourceDriver *self)
+{
+  gint rc;
+  if ((rc = MQTTClient_subscribe(self->client, self->topic,
+                                 mqtt_client_options_get_qos(&self->option))) != MQTTCLIENT_SUCCESS)
+    {
+      msg_error("mqtt: Error while subscribing to topic",
+                evt_tag_str("topic", self->topic),
+                evt_tag_int("qos", mqtt_client_options_get_qos(&self->option)),
+                evt_tag_str("error code", MQTTClient_strerror(rc)),
+                evt_tag_str("driver", self->super.super.super.super.id),
+                log_pipe_location_tag(&self->super.super.super.super.super));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_thread_init(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+  _client_init(self);
+}
+
+static void
+_thread_deinit(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+
+  MQTTClient_destroy(&self->client);
+}
+
+static gboolean
+_connect(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+
+  gint rc;
+
+  MQTTClient_connectOptions conn_opts;
+  mqtt_client_options_to_mqtt_connection_option(&self->option, &conn_opts);
+
+  if ((rc = MQTTClient_connect(&self->client, &conn_opts)) != MQTTCLIENT_SUCCESS)
+    {
+      msg_error("Error connecting mqtt client",
+                evt_tag_str("error code", MQTTClient_strerror(rc)),
+                evt_tag_str("client_id", mqtt_client_options_get_client_id(&self->option)),
+                log_pipe_location_tag(&self->super.super.super.super.super));
+      return FALSE;
+    }
+
+  if (!_subscribe_topic(self))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gboolean
+_unscribe_topic(MQTTSourceDriver *self)
+{
+  gint rc;
+  if ((rc = MQTTClient_unsubscribe(self->client, self->topic)) != MQTTCLIENT_SUCCESS)
+    {
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_disconnect(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+
+  _unscribe_topic(self);
+
+  MQTTClient_disconnect(self->client, MQTT_DISCONNECT_TIMEOUT);
+}
+
+static LogThreadedFetchResult
+_fetch(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+  ThreadedFetchResult result = THREADED_FETCH_ERROR;
+  char *topicName = NULL;
+  int topicLen;
+  MQTTClient_message *message = NULL;
+  LogMessage *msg = NULL;
+  gint rc = MQTTClient_receive(self->client, &topicName, &topicLen, &message, RECIVE_TIMEOUT);
```

Comment:
According to the doc[^1], `MQTTClient_receive()`'s result codes are a bit more complicated.

Success (return `THREADED_FETCH_SUCCESS`):
  1. `MQTTCLIENT_SUCCESS` and non-NULL `message`
  2. `MQTTCLIENT_TOPICNAME_TRUNCATED` and non-NULL `message`

Timeout (return `THREADED_FETCH_ERROR` maybe?):
  1. `MQTTCLIENT_SUCCESS` and NULL `message`

Error (return `THREADED_FETCH_ERROR` and log `rc`):
  1. Everything else

[^1]: https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/_m_q_t_t_client_8h.html#a4c2df88d00a3dadd510a8cb774739366

Reply:
Also, I think we should not use the `timeout` field, or use it with a smaller value, because our no-data mechanism does exactly the same. With that, you could add a `follow-freq()` option, which is the syslog-ng way of configuring polling.

Reply:
thanks!
commit: 4ad214aa5fb19a3e479688655a703ddf2cc6b93a